### DISCUSSION TITLE
(Go) Decoding into time.Duration

### DISCUSSION THREAD

**Discussion by curio77** [cue-lang/cue#1683] [Q&A] [answered]
Has anybody got a working example of how to define `time.Duration`-constrained (string) values in CUE and having them decoded to `time.Duration` in Go via `cue.Value.Decode()`?

I keep getting `cannot use value \"24h\" (type string) as int` as an error doing so (calling `Decode()` on a `Value` obtained from unifying a “constraints” value with a value parsed from YAML), passing a pointer to a struct that has the corresponding field typed `time.Duration`.

I.e., this CUE as constraint:

```cue
import "time"

D: time.Duration
```

and this YAML as value:

```yaml
D: "24h"
```

into this Go struct:

```go
type Into struct {
    D time.Duration
}
```

Edit: As should be expected, the same happens when attempting to decode a `Value` from this CUE:

```cue
import "time"

D: time.Duration
D: "24h"
```

straight away into the above Go struct.

**Comment by myitcv** [comment, marked as answer]
The behaviour of `cue.Value.Decode()` is best described by the documentation for [`cue.Value.Encode()`](https://pkg.go.dev/cuelang.org/go/cue#Context.Encode). 

What you are seeing doesn't actually have anything to do with CUE having explicit support for `time.Time`, but rather the fact that [`time.Time`](https://pkg.go.dev/time#Time) implements [`json.Marshaler`](https://pkg.go.dev/encoding/json#Marshaler) and [`json.Unmarshaler`](https://pkg.go.dev/encoding/json#Unmarshaler). `time.Duration` does not.

I suspect this is because there is not the same level of standardisation around durations as there is times/dates. 

cc @mvdan for more context/expertise.

**Reply by mvdan** [reply]
Indeed, what Paul said. Go does have its own text form for https://pkg.go.dev/time#Duration, via `String` and `ParseDuration`, but the fact that there's no `UnmarshalText` or `UnmarshalJSON` tells you that they aren't really meant as stable or standard encoding forms.

You could always wrap `time.Duration` into a type of your own that implements https://pkg.go.dev/encoding#TextUnmarshaler via `ParseDuration`, and presumably that should then work with CUE.

**Reply by mpvl** [reply]
A bit of brainstorming here:  durations are actually quite common and useful in configurations. We could consider adopt it as a pseudo standard for CUE and allow such parsing. To provide consistency, we should probably copy the parser and freeze its behavior in that case (as we should do for the time parse as well anyway). That way a relaxation of the parsing of durations would coincide with newer versions of CUE.

In general, I think it is okay to consider special handling in the decoder for types that are defined in the standard library.

**Reply by mpvl** [reply]
As another brainstorming thought: we could make such interpretation contingent on the presence of a `time.Duration` conjunct.
